src/policy/mallocspace/metadata.rs

```diff
@@ -1,61 +1,137 @@
 /// Thus, mark-bit is vm-dependant and is part of each VM's ObjectModel.
 ///
-#[cfg(target_pointer_width = "32")]
 pub(crate) const ALLOC_SIDE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
     is_global: false,
-    offset: 0,
+    offset: if cfg!(target_pointer_width = "64") {
+        LOCAL_SIDE_METADATA_BASE_ADDRESS.as_usize()
+    } else {
+        0
+    },
     log_num_of_bits: 0,
     log_min_obj_size: constants::LOG_MIN_OBJECT_SIZE as usize,
 };
 
+/// Metadata spec for the active page byte
+///
+/// We use a byte instead of a bit to avoid synchronization costs, i.e. to avoid
+/// the case where two threads try to update different bits in the same byte at
+/// the same time
 #[cfg(target_pointer_width = "64")]
-pub(crate) const ALLOC_SIDE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
+pub(crate) const ACTIVE_PAGE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
     is_global: false,
-    offset: LOCAL_SIDE_METADATA_BASE_ADDRESS.as_usize(),
-    log_num_of_bits: 0,
-    log_min_obj_size: constants::LOG_MIN_OBJECT_SIZE as usize,
+    offset: ALLOC_SIDE_METADATA_SPEC.offset
+        + metadata_address_range_size(&ALLOC_SIDE_METADATA_SPEC),
+    log_num_of_bits: 3,
+    log_min_obj_size: constants::LOG_BYTES_IN_PAGE as usize,
+};
+
+#[cfg(target_pointer_width = "32")]
+pub(crate) const ACTIVE_PAGE_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
+    is_global: false,
+    offset: ALLOC_SIDE_METADATA_SPEC.offset
+        + metadata_bytes_per_chunk(
+            ALLOC_SIDE_METADATA_SPEC.log_min_obj_size,
+            ALLOC_SIDE_METADATA_SPEC.log_num_of_bits,
+        ),
+    log_num_of_bits: 3,
+    log_min_obj_size: constants::LOG_BYTES_IN_PAGE as usize,
 };
 
 pub fn is_meta_space_mapped(address: Address) -> bool {
     let chunk_start = conversions::chunk_align_down(address);
-    ACTIVE_CHUNKS.read().unwrap().contains(&chunk_start)
+    is_chunk_mapped(chunk_start) && is_chunk_marked(chunk_start)
+}
+
+fn map_chunk_mark_space(chunk_start: Address) {
+    // We eagerly map 16Gb worth of space for the chunk mark bytes on 64-bits
+    #[cfg(target_pointer_width = "64")]
+    let start = chunk_start.saturating_sub(2048 * BYTES_IN_CHUNK);
+    #[cfg(target_pointer_width = "64")]
+    let size = 4096 * BYTES_IN_CHUNK;
+
+    // We eagerly map 2Gb (i.e. half the address space) worth of space for the chunk mark bytes on 32-bits
+    #[cfg(target_pointer_width = "32")]
+    let start = chunk_start.saturating_sub(256 * BYTES_IN_CHUNK);
+    #[cfg(target_pointer_width = "32")]
+    let size = 512 * BYTES_IN_CHUNK;
+
+    info!(
+        "chunk_start = {} mapping space for {} -> {}",
+        chunk_start,
+        start,
+        chunk_start + (size / 2)
+    );
+
+    if CHUNK_METADATA.try_map_metadata_space(start, size).is_err() {
+        panic!("failed to mmap meta memory");
+    }
 }
 
 pub fn map_meta_space_for_chunk(metadata: &SideMetadataContext, chunk_start: Address) {
```

Comment:
Comment,
e.g. a short list of the map ops this function does

Reply:
Not sure of what you mean by "map ops"?

Reply:
I mean the metadata areas you mmap in this function